proposals/public-suffix.md

```diff
@@ -0,0 +1,555 @@
+
+  //
+  // Gets the PSL dataset version if available
+  //
+  export function getVersion(): string?;
+}
+```
+
+### Behaviours
+
+#### 1. Private Suffixes
+
+By default, the lookup performed by `getRegistrableDomain()` should **include** all
+suffixes in the PSL dataset, i.e. both ICANN and Private (non-ICANN) suffixes.
+
+However, if an `options` object is passed to `getRegistrableDomain()` with key
+`excludePrivateSuffixes` set to `true`, then Private (non-ICANN) suffixes should be
+**excluded** from the lookup algorithm.
+
+##### Example
+
+`domain` = foo.bar.wixsite.com
+
+| Option                                   |   Registrable Domain | PSL Section |
+|------------------------------------------|:--------------------:|:-----------:|
+| excludePrivateSuffixes = false (default) |      bar.wixsite.com | Private     |
+| excludePrivateSuffixes = true            |          wixsite.com | ICANN       |
+
+#### 2. Multiple Suffixes per Domain
+
+The lookup performed by `getRegistrableDomain()` should select the **longest** matching
+suffix (unless specifically excluded using the `excludePrivateSuffixes` option).
+
+##### Example
+
+`domain` = foo.bar.lib.de.us
+
+| Candidate Suffix | PSL Section |
+|-----------------:|:-----------:|
+|            de.us | ICANN       |
+|        lib.de.us | Private     |
+
+The longest is lib.de.us, so `getRegistrableDomain()` resolves to bar.lib.de.us
+
+#### 3. PSL Special Rules
+
+The lookup performed by `getRegistrableDomain()` should adhere to the
+[PSL algorithm](https://github.com/publicsuffix/list/wiki/Format#formal-algorithm).
+In particular, it should apply the 'wildcard' and 'exception' rules in the PSL.
+
+##### Examples
+
+| Domain                   | Public suffix | Matched PSL rule | Explanation    | Registrable Domain |
+|-------------------------:|--------------:|-----------------:|:--------------:|-------------------:|
+| sub.domain.com           | com           | com              | Simple rule    | domain.com         |
+| sub.domain.co.uk         | co.uk         | co.uk            | Simple rule    | domain.co.uk       |
+| sub.domain.gov.ck        | gov.ck        | *.ck             | Wildcard rule  | domain.gov.ck      |
+| sub.domain.any.ck        | any.ck        | *.ck             | Wildcard rule  | domain.any.ck      |
+| sub.sub.domain.any.ck    | any.ck        | *.ck             | Wildcard rule  | domain.any.ck      |
+| www.ck                   | ck            | !www.ck          | Exception rule | www.ck             |
+| sub.www.ck               | ck            | !www.ck          | Exception rule | www.ck             |
+| sub.sub.www.ck           | ck            | !www.ck          | Exception rule | www.ck             |
+
+#### 4. Unknown Suffixes
+
+If no matching suffix is found in the PSL for a `domain` parameter, then unless it is determined
+to be specifically [invalid](#6-invalid-domain-parameter), it should be assumed the domain has a
+single-label suffix.
```

Comment:
I am happy to remove this assumption, and instead add an option to the API allowing users to opt-in to this behaviour of assuming a single-label suffix. I was guided in my analysis by the following:

- Google Chrome's Public Suffix implementation makes this assumption, I believe. See [GetDomainAndRegistry](https://source.chromium.org/chromium/chromium/src/+/main:net/base/registry_controlled_domains/registry_controlled_domain.h;l=184;drc=b3929870e1504f3937609eff00b1745fbd38916d): "If no matching rule is found in the effective-TLD data (or in the default data, if the resource failed to load), the last subcomponent of the host is assumed to be the registry."
- [tltds](https://github.com/remusao/tldts) also makes this assumption in its implementation. (tltds is a javascript library for obtaining Public Suffixes that is used by popular extensions such as [bitwarden password manager](https://addons.mozilla.org/en-US/firefox/addon/bitwarden-password-manager/) and [violentmonkey](https://addons.mozilla.org/en-US/firefox/addon/violentmonkey/).) However, there is also an open [tltds issue](https://github.com/remusao/tldts/issues/2121) caused by this assumption.
- Use Cases 1 (Filter Requests by Organization) and 3 (Detect Third-Party Requests) in this proposal are performance-sensitive, because they obtain the registrable domain for every request (not just top-level pages) as the user browses. I wanted to reduce the chance of making users' browsers feel permanently slower after installing these sorts of extensions. This assumption of a single-label suffix by default may allow a faster implementation due to avoiding an explicit PSL lookup for all single-label TLDs.
- Use Case 2 (Group Domains in UI) in this proposal may benefit from this assumption, because without it all unknown-suffixed domains would be grouped together.

Reply:
Hey, I'm the maintainer of `tldts` and wanted to clarify that the [issue linked](https://github.com/remusao/tldts/issues/2121) is about `tldts-experimental`, a faster but less accurate implementation of `tldts`. The main `tldts` package (which most projects are using) is fully accurate, and behaves in a slightly different way by returning a single label in case no rule was found matching the last label (or anything more specific), but will also return two flags: `isIcann` and `isPrivate`, which allow to know where the matching rule came from (i.e. the ICANN  or PRIVATE section of the public suffix list). If both of these flags are `false`, then we can assume that the fallback behavior of using the last label as public suffix was triggered. For example:
```sh
$ npx tldts https://example.thissuffixdoesnotexist
{
  "domain": "example.thissuffixdoesnotexist",
  "domainWithoutSuffix": "example",
  "hostname": "example.thissuffixdoesnotexist",
  "isIcann": false,
  "isIp": false,
  "isPrivate": false,
  "publicSuffix": "thissuffixdoesnotexist",
  "subdomain": ""
}
```

I lack context on this spec but this might allow to combine the single-label suffix fallback behavior with the use-cases outlined by @Rob--W above.

Reply:
The main use-cases for the PSL are `isValidUrl(...)` which does not want this behavior (because you want to know if the TLD is known) and `isSameSite(...)` for which it *almost* doesn't matter.

It only helps if you have made up TLDs. `printer.homenet` and `backup.homenet` would both have an empty eTLD and eTLD+1 would be "homenet". If you assume the last label must be a TLD then you end up with the correct behavior.